Georchestra pre authZN integration without direct LDAP access

groldan opened the pull request:
* Depends on https://github.com/georchestra/georchestra/pull/3497

Work split into two separate modules:
- `georchestra-integration/externalized-accounts`
- `georchestra-integration/georchestra-authnz`

# GeoNetwork externalized authorization framework

>  `georchestra-integration/externalized-accounts`

This module implements a GeoNetwork plugin that allows to externalize the
source of authoritative accounts, so that GeoNetwork `User`, `Group` domain objects,
and their relationships, are governed by an external provider in a reliable way.

No assumption about the origin of such authorities is made by this module, 
and hence it provides the services to keep internal GeoNetwork `User`s and `Group`s
synchronized with the authoritative source through a single abstraction that
need to be implemented for each particular case: a `CanonicalAccountsRepository`.

The premises under which these services work are the following:

- The external source of authorization MUST be able of handling the identity/identifier
dichotomy for `User` and `Group` entities. Whereas the identity may change (e.g. `User`
and `Group` name), the identifier may not (`User` and `Group` id).

- The external source of authorization MUST also provide, for each canonical `User` and
`Group`, a version-like property, that's used as a quick way to check whether GeoNetwork's
internal version of a given entity is up to date with the canonical, external system's.
The implementation details for this property are left to the provider implementation,
and can be given in the form of a hash string (e.g. hex-encoded SHA1), a timestamp, etc.

- Since GeoNetwork `User` and `Group` ids are internal and assigned by Hibernate,
this module implements the necessary means to keep internal entities in sync with
canonical `User`s and `Group`s.

## Implementing an external Authorization provider

In order to implement a GeoNetwork plugin that yields handling of users and groups
to an external authoritative source, the only requirement is to provide an implementation
of the following interface:

`org.geonetwork.security.external.repository.CanonicalAccountsRepository`

Optionally, several aspects of the internal workings can be controlled by providing
an instance of the `ExternalizedSecurityConfiguration` class, which controls
whether the groups are synchronized with "Organization" or authorization "Roles",
among other things.

### Extension point

This is the signature of the domain repository interface that needs to be implemented:

```java
public interface CanonicalAccountsRepository {
    List<CanonicalUser> findAllUsers();
    List<CanonicalGroup> findAllGroups();
    List<CanonicalGroup> findAllRoles();
    Optional<CanonicalUser> findUserByUsername(String username);
    Optional<CanonicalGroup> findGroupByName(String name);
    Optional<CanonicalGroup> findRoleByName(String name);
}
```

Note both "groups" and "roles" are represented as `CanonicalGroup`. Whether
`findAllGroups()` or `findAllRoles()` will be called by the synchronization
process is a matter of configuration, and defaults to "organizations" mode.

The `CanonicalUser` and `CanonicalGroup` abstractions are used as the authoritative
representations GeoNetwork internal `User` and `Group` object model instances
will be synchronized against.

```
public interface CanonicalUser {
    String getId(); // identity (immutable)
    String getUsername(); //identifier (mutable)
    String getFirstName();
    String getLastName();
    String getOrganization();
    String getLastUpdated();
    String getEmail();
    String getTitle();
    List<String> getRoles();
}
public interface CanonicalGroup {
    String getId(); // identity (immutable)
    String getName(); // identifier (mutable)
    GroupSyncMode getOrigin();
    String getDescription();
    String getLinkage();
    String getLastUpdated();
}
```

### Synchronization mechanism

An `org.geonetwork.security.external.integration.AccountsReconcilingService` 
instance will be contributed to the Spring application context, which has 
the following signature:

```java
public class AccountsReconcilingService {
    public Optional<User> findUpToDateUser(CanonicalUser canonical);
    public User forceMatchingGeonetworkUser(CanonicalUser canonicalUser);
    public void synchronize();
}
```

The `synchronize()` method performs the synchronization of internal GeoNetwork
users and groups with the authoritative versions provided by
`CanonicalAccountsRepository`.

Although it can be triggered manually if needed, it's primary client is
the `ScheduledAccountsSynchronizationService` service, which will
periodically call it based on its configuration (defaults to one minute
after each run, with a 10 seconds initial delay once the application started up).

## A word on Authentication

This module does not care about authentication. 

Authentication is a separate, orthogonal concern to Authorization.

That said, what this module provides is a prescribed way to obtain the user
details once the authentication has been performed, and to ensure the user
information is up to date with its canonical form at the same time.

Whatever the authentication mechanism, the process is to first obtain the
authenticated user's `CanonicalUser`, and then call `findUpToDateUser()`.

This method will return the GeoNetwork `User` only if it exists in the PostgreSQL
database, AND it's information is up to date with the canonical version (by means
of checking both representations' `lastUpdated` property for equality).

In the event that no up-to-date GeoNetwork `User` is returned, then
`forceMatchingGeonetworkUser()` will ensure to return an up-to-date `User`,
synchronizing it with the canonical representation if need be.

### Configuration

The following class contains all the configuration options for this subsystem,
so that the implementor can easily bind it to a `.properties` configuration file:

```
org.geonetwork.security.external.configuration.ExternalizedSecurityConfiguration
```

```
syncMode=orgs|roles
syncRolesFilter=.*
profiles.default=RegisteredUser
profiles.rolemappings.[GN_ADMIN]=Administrator
profiles.rolemappings.[GN_REVIEWER]=Reviewer
profiles.rolemappings.[GN_EDITOR]=Editor
profiles.rolemappings.[GN_USER]=RegisteredUser
scheduled.enabled=true
geonetwork.scheduled.timeUnit = SECONDS # MILLISECONDS/SECONDS/MINUTES/HOURS
scheduled.retryOnFailure = true
scheduled.initialDelay = 5
scheduled.retryDelay = 5
scheduled.delayBetweenRuns = 30
```

# Integrating GeoNetwork as a native Georchestra service

>  `georchestra-integration/georchestra-authnz`

## GeoNetwork spring configuration approach

GeoNetwork's approach to setting up the Spring application context is rather
involved. Here's a recap of how it's set up:

Traditional servlet/spring integration starts with setting up `springSecurityFilterChain`
servlet filter in `WEB-INF/web.xml`.

In this case it's a `jeeves.config.springutil.JeevesDelegatingFilterProxy`,
which will set up the application context to be loaded from a number of xml files
with different look up strategies:

* `src/main/webResources/WEB-INF/config-spring-geonetwork.xml`
* `src/main/webapp/WEB-INF/config-spring-geonetwork-parent.xml`

Instead of scattering georchestra specific changes all over the place inside 
`web/src/main/webapp/WEB-INF/` and/or `web/src/main/webResources/WEB-INF/`,
we should ideally be able to make almost no changes to those files (and hence
keep the web app module as vanilla as possible), since there are over 30 spring config
files over there already and it's quite confusing.

### Infrastructure configuration

Internal application business spring beans will be configured from a 
`config-spring-geonetwork.xml` config file at the classpath root, requiring
no further changes to any original geonetwork config file.

### Authentication entry point configuration

Authentication specific configuratio is to be placed in a `config-security-georchestra.xml`
file, and should contain no application business bean definitions.

This requires the following customization to
`webapp/WEB-INF/config-security/config-security.xml` though:

```xml
<import resource="classpath:config-security-georchestra.xml"/>
```

groldan committed:
Georchestra pre authZN integration without direct LDAP access

REVISIT: try to load role mappings from geonetwork.properties not working

REVISIT: Move JPAUserLink to gn-domain

    Couldn't make JPA load the JPAUserLink entity from a separate .jar.
    Ideally this entity should be loaded at run/startup time and contributed
    to the entity manager.

commit e1e5d12dcc98b8f6e72bcb3ec12d604195e9bb46
Author: Gabriel Roldan <gabriel.roldan@gmail.com>
Date:   Thu Aug 12 20:40:36 2021 -0300

WIP integrating with GeoNetwork, having issues with JPA

- security-proxy sends all the pre-authZN info, see docker/config/security-proxy/headers-mapping.properties
- GeorchestraPreAuthenticationFilter takes GeorchestraUserDetails and
matches a GN User (creating or updating it as necessary)
- All the LDAP configuration is removed, Georchestra integration shall
work against a higher level abstraction. Initial groups
synchronization will be done against an interface rather than directly
against LDAP, whose implementation should be a REST API (probably in
console)

commit f423d812864298141fa8203b4b033cf7e49a0aaa
Author: Gabriel Roldan <gabriel.roldan@gmail.com>
Date:   Fri Aug 6 23:08:35 2021 -0300

Exploratory prototype for pre-authZN

New module `georchestra-authn-authz` that will take care of
receiving all the necessary user information (properties and
credentials) from security-proxy and keep the GeoNetwork users in-sync
without cron jobs.

groldan committed:
Adapt to split of GeorchestraUser/GeorchestraUserDetails

groldan committed:
Use provided utility to compute a geor user hash if not provided

groldan committed:
Synchronize users and groups against georchestra's console REST API

pmauduit commented on georchestra-authnz/GEORCHESTRA_INTEGRATION.md:
```diff
@@ -0,0 +1,164 @@
+
+- `security-proxy` **must** be able to provide a unique user identifier,
+`sec-username` is not good engouh as it can be changed. We're not sure yet
+whether a stable UUID will be provided by LDAP? Mitigation strategy: by getting
```
we now are

pmauduit commented on georchestra-authnz/src/main/java/org/georchestra/geonetwork/logging/Logging.java:
```diff
@@ -0,0 +1,92 @@
+/**
+ * Logger abstraction using {@link String#format} syntax
+ */
+public class Logging {
```
What is the purpose of this class ? I thought you were using lombok and the slf4j-like annotations

groldan replied:
Can't use lombok in geonetwork, it just won't work due to other stuff I don't quite remember, related to JPA I think

groldan replied:
it's gone.

pmauduit commented on georchestra-authnz/src/main/java/org/georchestra/geonetwork/security/authentication/GeorchestraPreAuthenticationFilter.java:
```diff
@@ -0,0 +1,90 @@
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.web.authentication.preauth.AbstractPreAuthenticatedProcessingFilter;
+
+public class GeorchestraPreAuthenticationFilter extends AbstractPreAuthenticatedProcessingFilter {
```
Doesn't it look like something that could come from your previous PR's on common / security-proxy-spring-integration / ldap-account-management ?

groldan replied:
Added an explanation:
```
/**
 * Pre auth filter that gets the credentials as a {@link GeorchestraUser} from a
 * {@link GeorchestraSecurityProxyAuthenticationFilter} using composition, and
 * makes sure GeoNetwork user matches the georchestra user info.
 * 
 * @see GeorchestraToGeonetworkUserReconcilingService
 */

```

pmauduit commented on georchestra-authnz/src/main/java/org/georchestra/geonetwork/security/integration/AccountsSynchroizationService.java:
```diff
@@ -0,0 +1,90 @@
+import org.springframework.stereotype.Service;
+
+@Service
+public class AccountsSynchroizationService implements ApplicationListener<ApplicationContextEvent> {
```
typo in the class Name `AccountsSynchro*n*izationService`

pmauduit commented on georchestra-authnz/src/main/java/org/georchestra/geonetwork/security/integration/AccountsSynchroizationService.java:
```diff
@@ -0,0 +1,90 @@
+
+    @Override
+    public void onApplicationEvent(ApplicationContextEvent event) {
+        if (event instanceof ContextStartedEvent) {
+            synchronize();
+            log.info("Users and groups synhchronized upon application startup.");
+            log.info("Setting up scheduled synchronization every 1 minute");
+            this.scheduledExecutor = Executors.newScheduledThreadPool(1);
+            this.scheduledExecutor.scheduleWithFixedDelay(this::synchronize, 1, 1, TimeUnit.MINUTES);
+        } else if (event instanceof ContextStoppedEvent) {
+            if (this.scheduledExecutor != null) {
+                this.scheduledExecutor.shutdownNow();
+            }
+        }
```
I was not aware a scheduled service could be launched this way, I was more of using quartz scheduler, defining beans in some XML files (not a critic, just a remark, I am ok with it).

groldan replied:
Didn't bother to go with quarks, since a longer term goal would be to get changes in an event stream as soon as they happen in console, instead of having to poll.

pmauduit commented on georchestra-authnz/src/main/java/org/georchestra/geonetwork/security/integration/GeorchestraToGeonetworkUserReconcilingService.java:
```diff
@@ -0,0 +1,362 @@
+@Service
+public class GeorchestraToGeonetworkUserReconcilingService {
+
+    private static final Logging log = Logging.getLogger("org.georchestra.geonetwork.security.integration");
```
so it is used in here, but I still don't get why not using one of the logging library that should be available into the classpath

groldan replied:
indirectly we do, but this isolated me of whether it's using log4g or slf4j and let me use `String.format()` syntax. not a fan still, we could probably remove it, it's just that by then I've been fighting the logging config until I understood how to set it up properly (bah, had to comment out the call to `LogUtils.refreshLogConfiguration();` in `Geonetwork.java`...)

pmauduit commented on georchestra-authnz/src/main/java/org/georchestra/geonetwork/security/repository/GroupLink.java:
```diff
@@ -0,0 +1,86 @@
+import org.georchestra.security.model.Organization;
+
+/**
+ * Entity that represents a link between a geOrchestra {@link Organization} and
```
sometimes people in the georchestra community prefer to use the roles from the ldap instead of the organizations (see PR on the LDAP, and the fact that we also might need to address this usecase)

groldan replied:
yup, now I understood the problem you've been explaining to me over and over... If you look at `AccountsSynchronizationService` now, there's `synchronizeGroupsWithOrganizations()` and `synchronizeGroupsWithRoles()`.
The former will be the default mode, and the second will be enabled in `geonetwork.properties`.
I still don't get the settings in `geonetwork.properties` loaded to the `Environment` though, stopped fighting that in order to be able of making progress.

groldan committed:
Multimodule refactoring and ability sync groups on roles.

Refactor georchestra integration as a multi-module project:

```
georchestra-integration/
  externalized-accounts/
  georchestra-authnz/
  georchestra-utils/
```

groldan committed:
Use georchestra-testcontainers for complete integration testing

groldan committed:
Update georchestra integration documentation and fix spring configuration

* Update READMEs

* Fix geonetwork's web/* spring configuration to defer to
georchestra-authnz's module spring xml config files.

groldan committed:
git-ignore json files created by build

groldan committed:
Ensure legacy authentication with just sec-proxy and sec-username works on geonetwork

Make sure it is still possible to send just the
sec-proxy and sec-username request headers to GeoNetwork,
and the georchestra-authnz integration will synchronize
the internal geonetwork user/groups with the canonicals
from console.

groldan committed:
Repurpose docker entry point wait-for-ldap as wait-for-console

landryb commented:
fwiw with this PR merged, building the `georchestra-gn4-4.x-dev` branch on debian buster with java8 fails here:
```
[ERROR] Failed to execute goal on project gn-externalized-accounts: Could not resolve dependencies for project org.geonetwork-opensource:gn-externalized-accounts:jar:4.0.6-georchestra-SNAPSHOT: Could not find artifact org.georchestra:georchestra-testcontainers:jar:21.0-SNAPSHOT -> [Help 1]
```
that's with
```
$mvn clean install -f geonetwork/pom.xml -DskipTests -Dmaven.test.skip=true
```
( and no, i dont have docker installed, nor plan to install it ;) )

groldan commented:
Hi @landryb , yes, I'm working on it.
Thing is now geonetwork depends on some libraries that are part of georchestra.
The idea is to build it similarly to how we build georchestra's geoserver, but I'm scratching my head with that and with fixing the datafeeder ci build. Same commands work on my machine but get a "the operation was cancelled" in github actions.

all in all, trying to get a sane build as soon as possible. Stay tuned! :)

groldan commented:
> ( and no, i dont have docker installed, nor plan to install it ;) )

sounds good, we should switch to [podman](https://podman.io/) anyways

fvanderbiest commented:
> Same commands work on my machine but get a "the operation was cancelled" in github actions.

build timeout ?

groldan commented:
shame on me, yes:
```
 jobs:
   build:
     runs-on: ubuntu-latest
-    timeout-minutes: 10
+    timeout-minutes: 30
```

landryb commented:
fwiw master still fails, once i had manually built testcontainers module
```
mvn clean install -f testcontainers/pom.xml -DskipTests -Dmaven.test.skip=true
```
i get this, and i dunno where this gn-georchestra-authnz module/jar is supposed to come from (i get that its in the gn submodule, in the `georchestra-integration/georchestra-authnz/` dir)
```
[ERROR] Failed to execute goal on project gn-georchestra-authnz: Could not resolve dependencies for project org.geonetwork-opensource:gn-georchestra-authnz:jar:4.0.6-georchestra-SNAPSHOT: Could not find artifact org.georchestra:security-proxy-spring-integration:jar:21.0-SNAPSHOT -> [Help 1]
```

are the maven bits all okay ? a missing dependency somewhere ?

landryb commented:
the buildbot biulding master also fails with similar msgs since the merge to master:
```
[ERROR] Failed to execute goal on project gn-externalized-accounts: Could not resolve dependencies for project org.geonetwork-opensource:gn-externalized-accounts:jar:4.0.6-georchestra-SNAPSHOT: Could not find artifact org.georchestra:georchestra-testcontainers:jar:21.0-SNAPSHOT -> [Help 1]
```
